File: 03.py
import time
from libs import *


def main():
    print("--- Running Day 3 ---")

    start = time.perf_counter()

    try:
        with open("03.txt", "r") as f:
            data = f.readlines()
    except Exception as e:
        raise ValueError("Unable to read file") from e

    priority = 0
    for line in data:
        line = line.strip()
        cmp1 = set(line[: int(len(line) / 2)])
        cmp2 = set(line[int(len(line) / 2) :])
        (dup,) = cmp1 & cmp2
        if dup.islower():
            priority += ord(dup) - 96
        else:
            priority += ord(dup) - 64 + 26

    print(f"Output Part 1: {priority}")

    priority = 0
    for i in range(0, len(data), 3):
        cmp1 = set(data[i].strip())
        cmp2 = set(data[i + 1].strip())
        cmp3 = set(data[i + 2].strip())
        (dup,) = cmp1 & cmp2 & cmp3
        if dup.islower():
            priority += ord(dup) - 96
        else:
            priority += ord(dup) - 64 + 26

    print(f"Output Part 2: {priority}")

    end = time.perf_counter()
    rtime = (end - start) * 1000  # sec -> ms
    print(f"Took {format_runtime(rtime)}\n")


if __name__ == "__main__":
    main()

File: 03.js
const fs = require('fs');
const { performance } = require('perf_hooks');
const { formatRuntime } = require('./libs');

async function readInput() {
  const input = await fs.promises.readFile('03.txt', 'utf-8');
  return input.split('\n');
}

async function puzzle() {
  console.log('--- Running Day 3 ---');

  const start = performance.now();

  const input = await readInput();

  let priority = 0;
  input.forEach((line) => {
    const entry = line.trim();
    const cmp1 = new Set(entry.slice(0, entry.length / 2));
    const cmp2 = new Set(entry.slice(entry.length / 2));
    const [dup] = new Set([...cmp1].filter((i) => cmp2.has(i)));
    if (dup == dup.toLowerCase()) {
      priority += dup.charCodeAt(0) - 96;
    } else {
      priority += dup.charCodeAt(0) - 64 + 26;
    }
  });

  console.log('Part 1: ' + priority);

  priority = 0;
  for (let i = 0; i < input.length; i += 3) {
    const cmp1 = new Set(input[i].trim());
    const cmp2 = new Set(input[i + 1].trim());
    const cmp3 = new Set(input[i + 2].trim());
    const [dup] = new Set(
      [...cmp1].filter((i) => cmp2.has(i)).filter((i) => cmp3.has(i))
    );
    if (dup == dup.toLowerCase()) {
      priority += dup.charCodeAt(0) - 96;
    } else {
      priority += dup.charCodeAt(0) - 64 + 26;
    }
  }

  console.log('Part 2: ' + priority);

  const end = performance.now();
  const rtime = end - start;
  console.log('Took:', formatRuntime(rtime));
}

puzzle();
